README
###################################################
#
#  RAIDA
#
#  [ Root (http://root.cern.ch)    implementation of
#    AIDA (http://aida.freehep.org)  ]  
#  
#  
###################################################

 
Introduction:
-------------

RAIDA is a ROOT implementation of AIDA (Abstract Interfaces for Data Analysis).
The motivation for the development of RAIDA was to offer the 
possibility to create and fill n-tuple or histograms using standard ROOT 
objects with AIDA. All ROOT objects created with AIDA are stored in a 
ROOT file. Since our main application of RAIDA is to create ROOT output to
be analysed using the ROOT program, the current version can not read in 
the root files created. Furthermore only objects, which exist within ROOT 
can be created. 

This version of RAIDA supports the following AIDA objects:

ITREE (create output-file, create and change directories)  
IHistogram1D, IHistogram2D, IHistogram3D (1D, 2D, and 3D binned histograms) 
ICloud1D, ICloud2D, ICloud3D (1D, 2D, and 3D unbinned histograms)
IProfile1D, IProfile2D (1D and 2D profile histogram)
IAxis (get information about axis of IHistogram, ICloud, and IProfile)
ITuple (creating and filling of n-tuples with different data types:
        int, short, long, float, double)

Additional Information:
----------------------

The AIDA histograms hold more information about the data than those from ROOT. 
RAIDA can write out these additional information by including additional 
histograms in the output file. By default the additional histograms are 
omitted. To tell RAIDA to write out these histograms, the environment variable 
"AIDA_HISTOGRAMS_IN_ROOT" has to be set to "1":

export AIDA_HISTOGRAMS_IN_ROOT=1

Debug Information:
-----------------

To supporting the usage of a debugger like GDB (DDD), the object file has to
have some additional information. By default these information is not 
written to the object file. This can be changed by setting an environment
variable:

export RAIDADEBUG=1

Verbose Mode:
------------

To get additional information written to the screen during the operation of
RAIDA, set the following environment variable:

export RAIDA_DEBUG_VERBOSE=1


Installation (with cmake):
-------------------------

 1st: edit the file BuildSetup.cmake according to your
      system setup.
 
 2nd: type the following commands:
                                                                                                                                                            
    $ mkdir build
    $ cd build
    $ cmake -C ../BuildSetup.cmake ..
    $ make install
 
 For more information on CMake for the ILCSoftware check:
 http://ilcsoft.desy.de/portal/general_documentation/index_eng.html
 
 CMake's official website:
 http://www.cmake.org


Installation (old makefiles): -> Deprecated, please use cmake!
--------------------------------------------------------------

Set the environment variables that describe your root installation, e.g.:

   export ROOTSYS=/opt/products/root/5.08.00

   export LD_LIBRARY_PATH=$ROOTSYS/lib:$LD_LIBRARY_PATH
   export PATH=$ROOTSYS/bin:$PATH

and then build the library:
   
  cd ./src
  gmake
  cd ..

this will create ./lib/libRAIDA.a


Linking against RAIDA:
---------------------

In order to build your program with RAIDA you need to use the proper 
include files and libraries, e.g. in your Makefile add:
 
  CPPFLAGS += `aida-config --include`
  LIBS += `aida-config --lib`

  LIBS += `root-config --libs`

( the first two lines should work for any AIDA installation)


and before calling 'make' set the proper environment, e.g.:

 export ROOTSYS=/opt/products/root/5.08.00

 export LD_LIBRARY_PATH=$ROOTSYS/lib:$LD_LIBRARY_PATH
 export PATH=$ROOTSYS/bin:$PATH

 export RAIDA_HOME=/afs/desy.de/user/g/gaede/ilcsoft/RAIDA/v00-00

 . $RAIDA_HOME/bin/aida-setup.sh





ChangeLog
2008-10-14 15:17  engels

	* CMakeLists.txt, RAIDAConfig.cmake.in, src/CMakeLists.txt: added
	  32 bit compatibility option

2008-05-22 10:37  engels

	* ChangeLog, release.notes: ------- v01-04-02 -------

2008-05-21 17:32  engels

	* CMakeLists.txt, RAIDAConfig.cmake.in: added definitions
	  -DUSE_AIDA -DUSE_RAIDA

2008-04-15 16:41  gaede

	* ChangeLog: tag v01-04-01

2008-04-15 16:40  gaede

	* CMakeLists.txt, release.notes: ------------ version v01-04-01
	  --------------------

2008-04-03 14:14  engels

	* CMakeLists.txt: added aida-config and aida-setup.sh INSTALL rule

2008-01-30 18:42  killenberg

	* ChangeLog, Doxyfile, include/AIDA/IHistogramFactory.h,
	  include/RAIDA/IHistogram1DROOT.h,
	  include/RAIDA/IHistogram2DROOT.h,
	  include/RAIDA/IHistogram3DROOT.h,
	  include/RAIDA/IHistogramFactoryROOT.h, include/RAIDA/ITreeROOT.h,
	  src/IHistogram1DROOT.cc, src/IHistogram2DROOT.cc,
	  src/IHistogram3DROOT.cc, src/IHistogramFactoryROOT.cc,
	  src/ITreeROOT.cc: - Added Doxyfile - Introduced memory management
	  of histograms in IHistorgamFactoryROOT - Removed memory leaks in
	  IHistgram1D, 2D and 3D - Improved AIDA compatibility:
	  IHistorgamFactoryROOT::destroy(IBaseHistogram * hist) is now
	  implemented

2008-01-22 10:01  engels

	* RAIDAConfig.cmake.in: set both variables INCLUDE_DIR and
	  INCLUDE_DIRS for compatibility

2008-01-07 15:10  engels

	* CMakeLists.txt: bug fix: using CMAKE_MODULE_PATH with multiple
	  paths wouldn't work

2007-11-26 16:54  engels

	* BuildSetup.cmake: updated versions for ilcsoft release v01-03

2007-11-23 16:26  engels

	* BuildSetup.cmake: updated for ilcsoft release v01-03

2007-11-16 11:41  engels

	* BuildSetup.cmake: updated path to CMakeModules

2007-11-05 19:06  engels

	* RAIDAConfig.cmake.in: INCLUDE_DIR -> INCLUDE_DIRS

2007-11-05 18:43  engels

	* CMakeLists.txt, RAIDAConfig.cmake.in: added changes for writing
	  build definitions to RAIDAConfig.cmake

2007-10-01 12:37  engels

	* bin/aida-config: bug fixed as proposed by Adrian Vogel on the
	  LC-Forum

2007-09-12 09:58  engels

	* CMakeLists.txt: replaced CMAKE_XXX_DIR with PROJECT_XXX_DIR

2007-07-18 08:53  engels

	* CMakeLists.txt: updated version

2007-07-13 18:37  gaede

	* ChangeLog, release.notes: ---- release v01-03 -------------

2007-07-13 18:36  gaede

	* BuildSetup.cmake: updated pathes

2007-07-13 16:13  engels

	* README: updated

2007-07-12 15:03  engels

	* README: updated

2007-07-06 15:25  engels

	* README: added cmake build instructions

2007-07-04 16:07  engels

	* CMakeLists.txt: added rpath for apple

2007-07-02 14:56  engels

	* BuildSetup.cmake, CMakeLists.txt: added BuildSetup.cmake

2007-06-22 14:26  engels

	* CMakeLists.txt, src/CMakeLists.txt: retabed files

2007-05-29 17:47  engels

	* CMakeLists.txt: updated version

2007-05-22 18:11  gaede

	* example.cc: bug fix: proper path for ntuple: "/N-Tuples"

2007-05-21 12:30  engels

	* src/CMakeLists.txt: bug fix: -pedantic flag crashed in sl4

2007-05-16 16:32  engels

	* CMakeLists.txt, src/CMakeLists.txt: moved dependencies code to a
	  macro

2007-05-11 10:19  engels

	* CMakeLists.txt, src/CMakeLists.txt: bug fix: build_with and
	  link_with were calling load_package with required arg

2007-04-25 15:03  engels

	* CMakeLists.txt, RAIDAConfig.cmake.in, cmake_uninstall.cmake.in,
	  src/CMakeLists.txt: added CMake files

2007-03-27 16:15  tkraemer

	* src/: IDataPointSetROOT.cc, IFunctionROOT.cc,
	  IModelFunctionROOT.cc, IRangeSetROOT.cc, ITupleFactoryROOT.cc:
	  Removed critical statements to get rid of all warnings.

2007-03-23 15:43  tkraemer

	* ChangeLog: New ChangeLog.

2007-03-23 15:41  tkraemer

	* include/RAIDA/: IHistogram1DROOT.h, IHistogram2DROOT.h,
	  IHistogram3DROOT.h, IProfile1DROOT.h, IProfile2DROOT.h: Add all
	  #include statements to ensure the save compilation also on other
	  LINUX installations outside of DESY. In addition removed all
	  #includes which are not required by the class definition.

2007-03-23 15:38  tkraemer

	* src/: ICloud1DROOT.cc, ICloud2DROOT.cc, ICloud3DROOT.cc,
	  ITupleROOT.cc, LeafPoint.cc: Added some #include statements to
	  ensure save compilation on other LINUX installations outside of
	  DESY.

2007-03-23 15:35  tkraemer

	* src/GNUmakefile: Debug information turned off.

2007-02-08 18:29  tkraemer

	* src/: IProfile1DROOT.cc, IProfile2DROOT.cc: removed bug in
	  IProfilexD::fill() method. The 2nd, 3rd coordinate was not used.

2007-02-08 17:41  tkraemer

	* ChangeLog: Updated version of the ChangeLog.

2007-02-08 17:37  tkraemer

	* mainroot.cc, src/mainroot.cc, src/mainroot.cc.axis,
	  src/mainroot.cc.exception, src/mainroot.cc.old2,
	  src/mainroot.cc.tree1, src/mainroot.cc.tuple1,
	  src/mainroot.cc.tuple2, src/mainroot.cc.tuple3,
	  src/mainroot.cc.xx: Removed files which are now longer used
	  building the library.

2007-02-08 17:10  tkraemer

	* src/: IProfile1DROOT.cc, IProfile2DROOT.cc: Correction of bug: in
	  Function IProfile1D::binHeight() and IProfile2D::binHeight() the
	  sum of the weights was calculated incorrect, returning the number
	  of entries. This problem was removed in all IProfile classes.
	  Also other functions using equivalent functionality were checked.

2007-01-26 15:01  tkraemer

	* example.cc: Included more comments into the example program to
	  help the reader to understand how to use RAIDA.

2007-01-25 17:55  tkraemer

	* example.cc: some more example code + documentation.

2007-01-25 15:33  tkraemer

	* src/ITreeFactoryROOT.cc: Create ITree in memory if storeName
	  =="".

2007-01-24 19:08  tkraemer

	* example.cc: Some more functionality in the example program.

2007-01-24 17:16  tkraemer

	* src/ITreeROOT.cc: removed 1 "cout" used for debuging.

2007-01-24 11:57  tkraemer

	* include/AIDA/ITree.h, include/RAIDA/ITreeROOT.h,
	  src/ITreeROOT.cc: new implementation of the "ITree::ls()" method.
	  This method will return the ROOT directory at the moment and not

mainroot.cc
#include <stdlib.h>
#include <time.h>
#include <RAIDA/LeafPoint.h>
#include <TH1F.h>
#include <TH1D.h>
#include <TDirectory.h>
#include <TFile.h>
#include <TTree.h>
#include <TAxis.h>

#include <iostream>
#include <string>
#include <vector>

#include <AIDA/IAnalysisFactory.h>
#include <AIDA/ITreeFactory.h>
#include <AIDA/ITree.h>
#include <AIDA/IHistogramFactory.h>
#include <RAIDA/IHistogram1DROOT.h>
#include <AIDA/IHistogram2D.h>
#include <AIDA/IHistogram3D.h>
#include <AIDA/IProfile1D.h>
#include <AIDA/IProfile2D.h>

#include <AIDA/ITupleFactory.h>
#include <AIDA/ITuple.h>
#include <AIDA/ICloud1D.h>
#include <AIDA/ICloud2D.h>
#include <AIDA/ICloud3D.h>
#include <AIDA/IAxis.h>

using namespace AIDA ;
using namespace std;

int main()
{
  IAnalysisFactory * myaida;
  ITreeFactory * mytreefactory;
  ITree * mytree;
  IHistogramFactory * myhistofactory;
  myaida = AIDA_createAnalysisFactory();
  mytreefactory = myaida->createTreeFactory();


  std::string storeName("aidaroot.root");
  std::string storeType("root");
  bool readOnly = false;
  bool createNew = false;
  std::string options("keine");
  mytree = mytreefactory->create(storeName,storeType,readOnly,createNew,options);

  ITupleFactory * mytuplefactory = myaida->createTupleFactory(*mytree);
  myhistofactory = myaida->createHistogramFactory(*mytree);

  mytree->mkdir("ich");
  mytree->mkdir("du");
  mytree->mkdir("er");
  IHistogram1D *myhisto = myhistofactory->createHistogram1D("/x1","hi1",15,0,100);
  ICloud3D *cl1, *cl2, *cl3, *cl4, *cl5;
  ICloud1D *cl10,*cl11;
  ICloud2D *cl20,*cl21;
  ICloud3D *cl30,*cl31;
  cl10 = myhistofactory->createCloud1D("/ich/cl10","mycl10",400);
  cl20 = myhistofactory->createCloud2D("/ich/cl20","mycl20",400);
  cl30 = myhistofactory->createCloud3D("/ich/cl30","mycl30",400);
  cl1 = myhistofactory->createCloud3D("/ich/cl1","mycl1",500);
  cl2 = myhistofactory->createCloud3D("/ich/cl2","mycl2",15);
  cl3 = myhistofactory->createCloud3D("/du/cl3","mycl3",500);
  cl4 = myhistofactory->createCloud3D("/er/cl4","mycl4",15);
  //  cl6 = myhistofactory->createCloud1D("/er/cl6");

  mytree->cd("er");
  cl5 = myhistofactory->createCloud3D("/ich/cl5","mycl5",500);

  IHistogram1D *myhistoo = myhistofactory->createHistogram1D("/ich/xxl","hi xxl",5,0,1);
  //  mytree->cd("/");
  ITuple *mytuple;
  ITuple *mytuple1;
  vector<string> names,types;
  names.push_back("myd");
  types.push_back("D");
  names.push_back("myf");
  types.push_back("F");
  names.push_back("myi");
  types.push_back("I");
  names.push_back("mys");
  types.push_back("S");
  names.push_back("myl");
  types.push_back("L");
  names.push_back("myc");
  types.push_back("char");
  names.push_back("myb");
  types.push_back("bool");
  mytuple = mytuplefactory->create("/ich/mynt","my first ntuple",
				   names,types,"");

  mytuple1 = mytuplefactory->create("/ich/myntx","my first ntuplex",
				    " D col1 , int col2 , col3 , col4, I col5, col6, F col7","");
  mytuple->fill(0,11.1);
  mytuple->fill(1,(float)-11.2);
  mytuple->fill(2,113);
  mytuple->fill(3,(short)114);
  mytuple->fill(4,(long)115);
  mytuple->fill(5,'b');
  mytuple->fill(6,true);
  mytuple->addRow();

    for (int i = 0; i<490; i++)
      {
		cl1->fill( (double)i,(double)i,(double)i  );
		cl10->fill( (double)i );
		cl20->fill( (double)i,(double)i );
		cl30->fill( (double)i,(double)i,(double)i );
	  	cl2->fill( (double)i,(double)i,(double)i  );
	  	cl3->fill( (double)i,(double)i,(double)i  );
	  	cl4->fill( (double)i,(double)i,(double)i  );
	  	cl5->fill( (double)i,(double)i,(double)i  );
      }

    //        cl3->scale(3.);
    //        cl4->scale(3.);

/// Schliessen der Datei. -----------------------------------------------------
  mytree->cd("/er");
  //  ICloud1D *cl6;
  //  ICloud2D *cl7;
  //  ICloud3D *cl8;
  //  cl6 = myhistofactory->createCloud1D("cl6");
  //  cl7 = myhistofactory->createCloud2D("cl7");
  //  cl8 = myhistofactory->createCloud3D("cl8");

  //  IHistogram1D *myhi1 = myhistofactory->createHistogram1D("t1",5,0,1);
  //  IHistogram1D *myhi2 = myhistofactory->createCopy("/er/copy1",*myhi1);

  //  IHistogram2D *myhi3 = myhistofactory->createHistogram2D("t2",5,0,1,5,0,1);
  //  IHistogram2D *myhi4 = myhistofactory->createCopy("/er/copy2",*myhi3);

  //  IHistogram3D *myhi5 = myhistofactory->createHistogram3D("t3",5,0,1,5,0,1,5,0,1);
  //  IHistogram3D *myhi6 = myhistofactory->createCopy("/er/copy3",*myhi5);

  //  IProfile1D *mypr1 = myhistofactory->createProfile1D("/er/pr1",5,0,1);
  //  IProfile1D *mypr2 = myhistofactory->createCopy("/er/prc1",*mypr1);

  //  IProfile2D *mypr3 = myhistofactory->createProfile2D("/er/pr2",5,0,1,5,0,1);
  //  IProfile2D *mypr4 = myhistofactory->createCopy("/er/prc2",*mypr3);

  //IProfile2D *mypr2 = myhistofactory->createProfile2D("/er/pr2",5,0,1,5,0,1,2,3);

  //  cl11 = myhistofactory->createCopy("cl11",*cl10);
  //  cl21 = myhistofactory->createCopy("cl21",*cl20);
  //  cl31 = myhistofactory->createCopy("cl31",*cl30);

// variables Binning:
  vector<double> be;
  be.push_back(0.);
  be.push_back(1.);
  be.push_back(3.);

  //  IHistogram1D *myhi10 = myhistofactory->createHistogram1D("/er/v1","hv1",be);
  //  IHistogram2D *myhi20 = myhistofactory->createHistogram2D("/er/v2","hv2",be,be);
  //  IHistogram3D *myhi30 = myhistofactory->createHistogram3D("/er/v3","hv3",be,be,be);

  //  IProfile1D *mypr1 = myhistofactory->createProfile1D("pr1","mypr1",be);
  //  IProfile2D *mypr2 = myhistofactory->createProfile2D("pr2","mypr2",be,be);

  //    cout << "Variables Binning: " << endl;
  //    for ( int i=0 ; i<=mypr1->axis().bins()+1;i++)
  //      cout << i << " " 
  //  	 << mypr1->axis().binLowerEdge(i) << " " 
  //  	 << mypr1->axis().binUpperEdge(i) << " "
  //  	 << mypr2->xAxis().binLowerEdge(i) << " " 
  //  	 << mypr2->xAxis().binUpperEdge(i) << " "
  //  	 << mypr2->yAxis().binLowerEdge(i) << " " 
  //  	 << mypr2->yAxis().binUpperEdge(i) << " "
  //	 << myhi30->xAxis().binLowerEdge(i) << " " 
  //	 << myhi30->xAxis().binUpperEdge(i) << " "
  //	 << myhi30->yAxis().binLowerEdge(i) << " " 
  //	 << myhi30->yAxis().binUpperEdge(i) << " "
  //	 << myhi30->zAxis().binLowerEdge(i) << " " 
  //	 << myhi30->zAxis().binUpperEdge(i) << " "
  //  	 << endl;

// Add ... 
  IHistogram3D *myhi1 = myhistofactory->createHistogram3D("t1",3,0,1,3,0,1,1,0,1);
  IHistogram3D *myddt = myhistofactory->createHistogram3D("myddt",3,0,1,3,0,1,3,0,1);
  IHistogram3D *myhi2 = myhistofactory->createHistogram3D("t2",3,0,1,3,0,1,1,0,1);
  IHistogram2D *myhii = myhistofactory->createHistogram2D("txy",3,0,1,3,0,1);
  myhi1->fill(-10.,-10.,0.5);
  myhi1->fill(-10.,0.1,0.5);
  myhi1->fill(-10,0.4,0.5);
  myhi1->fill(-10.,0.7,0.5);
  myhi1->fill(-10.,10.,0.5);
  myhi1->fill(0.1,-10.,0.5);
  myhi1->fill(0.1,0.1,0.5);
  myhi1->fill(0.1,0.4,0.5);
  myhi1->fill(0.1,0.7,0.5);
  myhi1->fill(0.1,10.,0.5);
  myhi1->fill(0.4,-10.,0.5);
  myhi1->fill(0.4,0.1,0.5);
  myhi1->fill(0.4,0.4,0.5);
  myhi1->fill(0.4,0.7,0.5);
  myhi1->fill(0.4,10.,0.5);
  myhi1->fill(0.7,-10.,0.5);
  myhi1->fill(0.7,0.1,0.5);
  myhi1->fill(0.7,0.4,0.5);
  myhi1->fill(0.7,0.7,0.5);
  myhi1->fill(0.7,10.,0.5);
  myhi1->fill(10.,-10.,0.5);
  myhi1->fill(10.,0.1,0.5);
  myhi1->fill(10.,0.4,0.5);
  myhi1->fill(10.,0.7,0.5);
  myhi1->fill(10.,10.,0.5);
  srand(time(NULL));
  cout << "### vor cast " << RAND_MAX << endl;
  for (int j=0;j<1000000;j++)
    {
      double dd = rand() % 100 ;
      double dda = 1.2*(double)(rand() % 100)/100. -0.1 ;
      double ddb = 1.2*(double)(rand() % 100)/100. -0.1 ;
      double ddc = 1.2*(double)(rand() % 100)/100. -0.1 ;
      //      cout << "// rnd: " << dd << " " << dda << " "  << ddb << " "  << endl; 
      myhisto->fill(dd);
      myhii->fill(dda,ddb);
      myddt->fill(dda,ddb,ddc);
    }
  dynamic_cast<IHistogram1DROOT*>(myhisto)->printContents();
  dynamic_cast<IHistogram2DROOT*>(myhii)->printContents();
  dynamic_cast<IHistogram3DROOT*>(myddt)->printContents();
  cout << "### nach print" << endl; 
  myhi2->fill(-20,-20,0.5);
  myhi2->fill(-20,0.3,0.5);
  myhi2->fill(-20,0.6,0.5);
  myhi2->fill(-20,0.9,0.5);
  myhi2->fill(-20,20.,0.5);
  myhi2->fill(0.3,-20,0.5);
  myhi2->fill(0.3,0.3,0.5);
  myhi2->fill(0.3,0.6,0.5);
  myhi2->fill(0.3,0.9,0.5);
  myhi2->fill(0.3,20,0.5);
  myhi2->fill(0.6,-20,0.5);
  myhi2->fill(0.6,0.3,0.5);
  myhi2->fill(0.6,0.6,0.5);
  myhi2->fill(0.6,0.9,0.5);
  myhi2->fill(0.6,20.,0.5);
  myhi2->fill(0.9,-20.,0.5);
  myhi2->fill(0.9,0.3,0.5);
  myhi2->fill(0.9,0.6,0.5);
  myhi2->fill(0.9,0.9,0.5);
  myhi2->fill(0.9,20.,0.5);
  myhi2->fill(20.,-20.,0.5);
  myhi2->fill(20.,0.3,0.5);
  myhi2->fill(20.,0.6,0.5);
  myhi2->fill(20.,0.9,0.5);
  myhi2->fill(20.,20.,0.5);
  // 2d-histo
  myhii->fill(-20,-20,0.5);
  myhii->fill(-20,0.3,0.5);
  myhii->fill(-20,0.6,0.5);
  myhii->fill(-20,0.9,0.5);
  myhii->fill(-20,20.,0.5);
  myhii->fill(0.3,-20,0.5);
  myhii->fill(0.3,0.3,0.3);
  myhii->fill(0.3,0.6,0.3);
  myhii->fill(0.3,0.9,0.3);
  myhii->fill(0.3,20,0.5);
  myhii->fill(0.6,-20,0.5);
  myhii->fill(0.6,0.3,0.6);
  myhii->fill(0.6,0.6,0.6);
  myhii->fill(0.6,0.9,0.6);
  myhii->fill(0.6,20.,0.5);
  myhii->fill(0.9,-20.,0.5);
  myhii->fill(0.9,0.3,0.9);
  myhii->fill(0.9,0.6,0.9);
  myhii->fill(0.9,0.9,0.9);
  myhii->fill(0.9,20.,0.5);
  myhii->fill(20.,-20.,0.5);
  myhii->fill(20.,0.3,0.5);
  myhii->fill(20.,0.6,0.5);
  myhii->fill(20.,0.9,0.5);
  myhii->fill(20.,20.,0.5);

  IHistogram3D *myhi3 = myhistofactory->add("added",*myhi1,*myhi2);
  // IHistogram1D *proj = myhistofactory->projectionY("pr",*myhii);  
  IHistogram1D *proj = myhistofactory->sliceY("pr",*myhii,1);  
  //  myhi1->add(*myhi2);
  //  IHistogram2D *myhi3 = myhi1; 
  cout << "### noch da! proj: " << proj << endl; 
  for (int i=0;i<=2;i++)
    {
  //        cout << proj->binMean(0) << " ";
      cout << proj->binHeight(i) << " ";
    }
  cout << endl; 
  for (int i=0;i<=2;i++)
    {
      for (int j=0;j<=2;j++)
	cout << myhii->binHeight(j,i) << " "; 
      cout << endl;
    }
  //cout << "### das ist nan" << endl;
  for (int i=-2;i<=2;i++)
    {
      for (int j=-2;j<=2;j++)
	cout << myhi3->binEntries(i,j,0) << " ";
      cout << endl;
    }
  for (int i=-2;i<=2;i++)
    {
      for (int j=-2;j<=2;j++)
	cout << myhi3->binMeanX(i,j,0) << " ";
      cout << endl;
    }
  for (int i=-2;i<=2;i++)
    {
      for (int j=-2;j<=2;j++)
	cout << myhi3->binMeanY(i,j,0) << " ";
      cout << endl;
    }

  IHistogram1D *myhi5 = myhistofactory->createHistogram1D("t3",3,0,1);
  IHistogram1D *myhi6 = myhistofactory->createHistogram1D("t4",3,0,1);
  myhi5->fill(-10.,2);
  myhi5->fill(-10.,2);
  myhi5->fill(0.1,4);
  myhi5->fill(0.4,6);
  myhi5->fill(0.7,8);
  myhi5->fill(10.,10);
  myhi5->fill(10.,10);
  myhi5->fill(10.,10);

  myhi6->fill(-10.,2);
  myhi6->fill(0.3,2);
  myhi6->fill(0.6,2);
  myhi6->fill(0.9,0);
  myhi6->fill(10.,0);

  //  IHistogram1D *myhi8 = myhistofactory->add("added",*myhi6,*myhi6);
  IHistogram1D *myhi7 = myhistofactory->divide("sub",*myhi5,*myhi6);
  //  IHistogram1D *myhi7 = myhistofactory->add("subadd",*myhi9,*myhi8);
  //  IHistogram1DROOT *myhi7 = dynamic_cast<IHistogram1DROOT*>(myhi5);
  //  myhi7->divide(*myhi6);


  for (int i=-3;i<=4;i++)
    cout << myhi5->binHeight(i) << " "; 
  cout << endl;
  for (int i=-3;i<=4;i++)
    cout << myhi5->binError(i) << " "; 
  cout << endl;
  for (int i=-3;i<=4;i++)
    cout << myhi5->binEntries(i) << " ";
  cout << endl;
  for (int i=-3;i<=4;i++)
    cout << myhi5->binMean(i) << " ";
  cout << endl;

  cout 
    << myhi5->coordToIndex(-20) << " " 
    << myhi5->coordToIndex(0.1) << " " 
    << myhi5->coordToIndex(0.5) << " " 
    << myhi5->coordToIndex(0.8) << " " 
    << myhi5->coordToIndex(30) << " " 
    << endl;

  //  mytree->cd("/");
  cout << "u and o bin: " << IAxis::UNDERFLOW_BIN << " " 
       << IAxis::OVERFLOW_BIN << endl;  


  TH1 * achshi = new TH1D("achstest","achstest",100,0.,1.);
  cout << "unten " << achshi->GetXaxis()->GetFirst() << " oben " 
       << achshi->GetXaxis()->GetLast() << endl;
  achshi->GetXaxis()->SetRange(23,56);
  cout << "unten " << achshi->GetXaxis()->GetFirst() << " oben " 
       << achshi->GetXaxis()->GetLast() << endl;

  IHistogram3D *my3d = myhistofactory->createHistogram3D(
      "u2",3,0.5,3.5,3,0.5,3.5,3,0.5,3.5);

  int l = 0 ;
  for (int i=0;i<5;i++)
    {
      for (int j=0;j<5;j++)
	{
	  for (int k=0;k<5;k++)
	    {
	      l++;
	      my3d->fill( (double)k, (double)j, (double)i, (double)l );
	      cout << i << " " << j << " " << k << " " << l << endl;
	    }
	}
    }

  for (int i=-2;i<3;i++)
    {
      for (int j=-2;j<3;j++)
	{
	  for (int k=-2;k<3;k++)
	    {
	      cout << my3d->binHeight(k,j,i) << " ";
	    }
	  cout << " L";
	}
      cout << endl << endl; 
    }

  IHistogram2D *mystueck = myhistofactory->sliceXY("stueck",*my3d,0,0);
  for (int j=-2;j<3;j++)
    {
      for (int i=-2;i<3;i++)
	{
	  //	  cout << mystueck->binHeight(j,i) << " ";
	  cout << mystueck->binEntries(j,i) << " ";
	}
      cout << endl ;
    }

//------------------------------------------------------------------------
        mytree->commit();
  ///      mytree->close();

}

